src/apiutil.h

```diff
@@ -308,3 +316,208 @@ bool hasInsufficientMaterial(Color c, const Position& pos) {
 
     return true;
 }
+
+enum FEN_VALIDATION : int {
+  FEN_MISSING_SPACE_DELIM = -12,
+  FEN_INVALID_NB_PARTS = -11,
+  FEN_INVALID_CHAR = -10,
+  FEN_FILE_SIZE_MISMATCH = -9,
+  FEN_INVALID_NUMBER_OF_RANKS = -8,
+  FEN_INVALID_POCKET_INFO = -7,
+  FEN_INVALID_SIDE_TO_MOVE = -6,
+  FEN_INVALID_CASTLING_INFO = -5,
+  FEN_INVALID_EN_PASSANT_SQ = -4,
+  FEN_INVALID_CHECK_COUNTER = -3,
+  FEN_INVALID_HALF_MOVE_COUNTER = -2,
+  FEN_INVALID_MOVE_COUTNER = -1,
+  FEN_EMPTY = 0,
+  FEN_OK = 1
+};
+
+
+FEN_VALIDATION validate_fen(const std::string& fen, const Variant* v) {
+
+  // 0) Layout
+  // check for empty fen
+  if (fen.size() == 0) {
+    std::cerr << "Fen is empty." << std::endl;
+    return FEN_EMPTY;
+  }
+
+  // check for space
+  if (fen.find(' ') == std::string::npos) {
+    std::cerr << "Fen misses space as delimiter." << std::endl;
+    return FEN_MISSING_SPACE_DELIM;
+  }
+
+  std::vector<std::string> fenParts;
+  std::vector<std::string> starFenParts;
+  std::stringstream ss(fen);
+  std::stringstream ss2(v->startFen);
+
+  std::string curPart;
+  while (std::getline(ss, curPart, ' '))
+    fenParts.emplace_back(curPart);
+  while (std::getline(ss2, curPart, ' '))
+    starFenParts.emplace_back(curPart);
+  const unsigned int nbFenParts = starFenParts.size();
+
+  // check for number of parts
+  if (fenParts.size() != nbFenParts) {
+    std::cerr << "Invalid number of fen parts. Expected: " << nbFenParts << " Actual: " << fenParts.size() << std::endl;
+    return FEN_INVALID_NB_PARTS;
+  }
+
+  // 1) Part
+  // check for valid characters
+  for (char c : fenParts[0])
+    if (!isdigit(c) && v->startFen.find(c) == std::string::npos) {
```

Comment:
I think this can break in some cases:
* `+` and `~` signs for promoted pieces that are not present in the starting position
* pieces that can only be obtained via promotion (e.g., a black queen in orda)

If you have not seen it already, I have similar logic in the variant config parser (to validate the start FEN), but it directly checks the array of piece types, which should be safer than relying on the start FEN. https://github.com/ianfab/Fairy-Stockfish/blob/369c55f9f2dce34b9cb65ba78ef1649543d11f59/src/parser.cpp#L336-L347

Reply:
I'll include `+`, `~` and `/` as valid characters and will use `v->pieceToChar` insteda.